Show version/language selectors below the title

humitos opened the pull request:
Allow users to decide whether or not showing a version and language selectors below the title of the documentation.

I'm adding two new configs here:

* `version_selector`
* `language_selector`

I'm re-using the code written some time ago in #438 for CSS styling.

**Ideally, we should include this in the 3.0.0 final version and release it on October 7th as planned**. That way, people using our own theme will get a nice and integrated version/language selector without the need of using `flyout_display=attached` which re-implements _all the flyout_ and requires disabling the floating one to avoid duplications.

### Example


![Peek 2024-09-17 16-30](https://github.com/user-attachments/assets/379ff382-5b48-4937-940a-5aa2ca497303)


* Live example: https://test-builds.readthedocs.io/en/ver-lang-selectors/

ericholscher commented on docs/configuring.rst:
```diff
@@ -192,6 +198,24 @@ Miscellaneous options
+.. confval:: version_selector
+
+    Display a version selector below the title.
+    This feature makes usage of `Read the Docs Addons <https://docs.readthedocs.io/page/addons.html>`_ for this,
+    so it's required the documentation to be hosted on Read the Docs.
```
I do wonder about adapting this so it can use a JSON file as well in a future version, but seems fine for now.

ericholscher replied:
```suggestion
    This feature makes usage of `Read the Docs Addons <https://docs.readthedocs.io/page/addons.html>`_ for this,
    so it's required the documentation to be hosted on Read the Docs.
    It only appears when there are more than 1 active version. 
```

ericholscher commented on docs/configuring.rst:
```diff
@@ -192,6 +198,24 @@ Miscellaneous options
+
+    Display a language selector below the title.
+    This feature makes usage of `Read the Docs Addons <https://docs.readthedocs.io/page/addons.html>`_ for this,
+    so it's required the documentation to be serverd on Read the Docs.
```
```suggestion
    so it's required the documentation to be served on Read the Docs.
    It only appears when there is more than 1 active language. 
```

ericholscher reviewed (commented):
I reviewed the Python & JS, which looks pretty solid. Not knowledgable enough on the CSS :)

agjohnson commented on sphinx_rtd_theme/static/js/versions.js_t:
```diff
@@ -123,3 +124,68 @@ function renderLanguages(config) {
+    </select>
+  `;
+
+  dropdownSelectors.insertAdjacentHTML("beforeend", versionSelect);
+  const versionSelector = document.querySelector("#readthedocs-version-selector");
+  versionSelector.addEventListener("change", onSelectorSwitch);
+  versionSelector.innerHTML = versionSelect;
```
I'm having a bit of trouble figuring out why some of this is needed. It feels like adding, querying, then adding the elements again could be simplified into something easier to follow.

The reason I hardcoded multiple inner divs in `div.switch-menus` is so that we already have an easy target for this HTML:

```suggestion
  versionSwitch.innerHTML = versionSelect;
  versionSwitch.firstChild.addEventListener("change", onSelectorSwitch);
```

humitos replied:
I followed this approach and it worked fine.

agjohnson commented on sphinx_rtd_theme/static/js/versions.js_t:
```diff
@@ -123,3 +124,68 @@ function renderLanguages(config) {
+    languages = languages.sort((a, b) => a.language.name.localeCompare(b.language.name));
+
+    const languageSelect= `
+      <select id="readthedocs-version-selector">
```
Same note on id usage here:

```suggestion
      <select>
```

CSS selector is `div.switch-menus > div.version-switch > select` and has plenty of specificity.

agjohnson commented on sphinx_rtd_theme/static/js/versions.js_t:
```diff
@@ -123,3 +124,68 @@ function renderLanguages(config) {
+       </select>
+    `;
+
+    dropdownSelectors.insertAdjacentHTML("beforeend", languageSelect);
+    const languageSelector = document.querySelector("#readthedocs-language-selector");
+    languageSelector.addEventListener("change", onSelectorSwitch);
+    languageSelector.innerHTML = languageSelect;
```
```suggestion
    switchLanguage.innerHTML = languageSelect;
    switchLanguage.firstChild.addEventListener("change", onSelectorSwitch);
```

agjohnson commented on src/sass/_theme_layout.sass:
```diff
@@ -231,10 +231,48 @@ html
+      &:hover, &:active, &:focus
+        background: rgba(255, 255, 255, .1)
+        color: rgba(255, 255, 255, .5)
+      &:after
```
It definitely needs something visual, it doesn't look like you are supposed to click it yet. This is where most frameworks recreate a custom select, like you're pointing at.

agjohnson commented on sphinx_rtd_theme/static/js/versions.js_t:
```diff
@@ -123,3 +124,68 @@ function renderLanguages(config) {
+          (version) => `
+          <option
+              value="${ version.slug }"
+              ${ config.versions.current.slug === version.slug ? 'selected="selected"' : '' }
```
A couple things this needs to support:

- Either don't show this menu on PR builds or show this menu with the PR build as the selected version, so that there isn't a conflicting menu on PR builds

![image](https://github.com/user-attachments/assets/5755042e-cb3a-492d-af51-41a319a5cb3f)

- Default to the default branch. Otherwise it just shows the first option (`0.5.0` in this case, which is not good UX)

humitos replied:
> show this menu with the PR build as the selected version

I followed this approach.

humitos commented on sphinx_rtd_theme/static/js/versions.js_t:
```diff
@@ -1,3 +1,4 @@
+{%- if theme_flyout_display == "attached" %}
```
I followed this pattern and I passed `prettier` on this file as well.

humitos commented:
I've addressed all the feedback received. This is how it looks now:

![Peek 2024-09-19 11-59](https://github.com/user-attachments/assets/3f5fd30e-0fec-42a0-bf7f-cc4570d4366e)

I think it's good enough to me. Let me know if you have any other feedback.

ericholscher commented:
This looks like a great start. I'd love to have some indication of what the strings are (eg. the branch and language icons next to them), but that can be added in the next iteration.

agjohnson commented:
This looks good. I have a lot of CSS fixes that would help this, it's probably easier to just show the change though.

The width and padding could use tuning, as the field width look unintentional stretched to the nav bar width

![image](https://github.com/user-attachments/assets/f8ccfb0e-39f3-4a5a-8543-84c9b4e6d3d0)

I wouldn't add icons here yet, I think I'd play around with making this a combination dropdown field instead of native select elements next. Making this into a custom dropdown gives us some room to explain default versions/languages, add icons for versions/prs/languages, and add more supporting context to the UI that is missing with just a `select`.

agjohnson commented:
@humitos how are you testing this locally?

humitos commented:
I'm not 😅. I'm just pushing and re building the test-builds version

agjohnson commented:
Roger. What would we need to do to make this locally reproducible? Remote testing seems okay for an initial test, but we don't want to hide these styles from other contributors or things could be prone to breaking

humitos commented:
Yeah, I didn't want to jump to deep in the mocking for this just yet.

I suppose we can follow the pattern we had for multiple versions on debugging but in this case fire the custom event with mocked data. Shouldn't be hard

agjohnson commented:
I threw up my changes at #1603 because of the issues with testing locally. It changes the fields to variable width instead of hardcoding them to be 10em.

humitos commented:
> Mock the custom event with placeholder data?

@agjohnson I've done this at https://github.com/readthedocs/sphinx_rtd_theme/pull/1606

humitos commented:
@agjohnson this PR is ready to reveiw and merge 👍🏼

agjohnson reviewed (approved):
Looks good for a test! I'm curious to see what this change does in the wild, I've wanted this for a while.

cristian64 commented:
With `display_version` removed, what's the workflow now for showing a single, static version in the navigation bar for docs that are _not_ hosted on Read the Docs?

humitos commented:
@cristian64 Hi, another user also asked this question and opened an issue at #1624. Please, subscribe there get updates on this topic. We will be working in a solution for this case.
